### ISSUE TITLE
Compiler fails on function with if-statement multiplied by float

### ISSUE THREAD

**Issue by Mokosha** [rust-lang/rust#32368]
I'm not sure if this is a bug or if it was reported before. I'm noticing the following behavior in the nightly version of rustc through http://play.rust-lang.org --

The following program correctly prints the output '6':

```
fn bug(x: f32) -> f32 {
    6.0 * if x > 0.0 {
        1.0
    } else {
        2.0
    }
}

fn main() {
    println!("{:?}", bug(0.5));
}
```

However, the following program fails to compile due to type mismatch:

```
fn bug(x: f32) -> f32 {
    if x > 0.0 {
        1.0
    } else {
        2.0
    } * 6.0
}

fn main() {
    println!("{:?}", bug(0.5));
}

//////////////////////////////////////////////////////////////////////////////////////////
<anon>:3:9: 3:12 error: mismatched types:
 expected `()`,
    found `_`
(expected (),
    found floating-point variable) [E0308]
<anon>:3         1.0
                 ^~~
<anon>:3:9: 3:12 help: see the detailed explanation for E0308
<anon>:5:9: 5:12 error: mismatched types:
 expected `()`,
    found `_`
(expected (),
    found floating-point variable) [E0308]
<anon>:5         2.0
                 ^~~
<anon>:5:9: 5:12 help: see the detailed explanation for E0308
<anon>:6:7: 6:12 error: type `_` cannot be dereferenced
<anon>:6     } * 6.0
               ^~~~~
error: aborting due to 3 previous errors
playpen: application terminated with error code 101
```

Shouldn't these be the same? Either both should fail or neither should fail?

**Comment by frewsxcv** [comment]
Maybe related to https://github.com/rust-lang/rust/issues/21222

**Comment by P1start** [comment]
This isn’t a typesystem error, it’s a quirk of the parser. The offending code is parsed as follows:

```rust
fn bug(x: f32) -> f32 {
    if x > 0.0 {
        1.0
    } else {
        2.0
    }
    *6.0
}
```

Any statement that begins with `if` (among other keywords) will be interpreted as a statement returning `()` (which is why you don’t have to put a semicolon after every `if` statement you don’t use as a value). The type errors are due to the fact that any `if` expression used as a statement has to return `()`, so the compiler doesn’t like the fact that both branches return `f64`s. The third error is because it interprets the `*6.0` as a new statement (because the previous `if` expression was interpreted as its own statement) and so thinks it is an attempt to dereference a floating-point number, which obviously makes no sense.

(This doesn’t effect the first testcase because the statement doesn’t start with `if`; thus, it doesn’t try to interpret the `if` expression as a full statement of its own.)

Because of this, I think this is intentional, but it would be quite nice to have a better error message for this sort of thing—it’s easy to get caught by, and tricky to figure out the problem. (This also applies to `-`, `<`, `(`, `[`, `&`, `&&`, `|`, and `||`, which can all both start an expression and extend one.)

**Comment by swiftcoder** [comment]
The case from #34620 that is missing in this issue is that the parser can deal with this anytime there are explicit statement terminators. i.e. in a let statement:

```rust
fn bug(x: f32) -> f32 {
    let r = if x > 0.0 {
        1.0
    } else {
        2.0
    } * 6.0;
    return r;
}

fn main() {
    println!("{:?}", bug(0.5));
}
```

This is unfortunate, as a legal expression in a let statement is illegal on its own...

**Comment by vlastachu** [comment]
I thought that in rust every `if` expression should ends with semicolon (if result not used). It looks ugly, of course, but it is have different semantic in functions last expressions.

In short, I was trying to fix it, but compilation aborts on `libcore/num/flt2dec/mod.rs:361`:

```
    if exp < 0 {
        parts[n] = Part::Copy(if upper { b"E-" } else { b"e-" });
        parts[n + 1] = Part::Num(-exp as u16);
    } else {
        parts[n] = Part::Copy(if upper { b"E" } else { b"e" });
        parts[n + 1] = Part::Num(exp as u16);
    }
    &parts[..n + 2]
```

And now it looks like that this behavior of `if` expression will never implemented. Btw user can call methods and get elements of array on `if` expression everywhere. And as mentioned @swiftcoder with `let ... = if...` user can construct any possible expression. So the single disallowed use case is binary operator with `if` expression on left side in the last expression of function. It looks inconsistent.

**Comment by oli-obk** [comment]
Update: the compiler diagnostic now informs the user on how to fix this

```
help: parentheses are required to parse this as an expression
  |
2 |     (if x > 0.0 {
3 |         1.0
4 |     } else {
5 |         2.0
6 |     }) * 6.0
  |
``` 

I'm going to close this issue as I think we need an RFC to change the semantics of such expressions
